docs/blog/2025-01-20-linking-ts-in-monorepos.md

```diff
@@ -0,0 +1,666 @@
+```
+
+This new `tsconfig.json` is the entry point for TypeScript project references, pointing to all individual TypeScript configs of the projects that are part of the monorepo workspace.
+
+```json {% fileName="tsconfig.json" %}
+{
+  "files": [],
+  "references": [{ "path": "./packages/lib-a" }, { "path": "./apps/myapp" }]
+}
+```
+
+This is distinct from `tsconfig.base.json`, which is used to share common configurations across the workspace:
+
+```diff {% fileName="tsconfig.base.json" %}
+{
+  "compilerOptions": {
+    "target": "ES2020",
+    "module": "NodeNext",
+    "strict": true,
+    "moduleResolution": "NodeNext",
++   "composite": true,
++   "declaration": true,
++   "declarationMap": true,
+    "sourceMap": true,
+    "baseUrl": ".",
+-   "rootDir": ".",
+    "paths": {
+      "@ts-monorepo-linking/lib-a": ["packages/lib-a/src/index.ts"]
+    }
+  }
+}
+```
+
+Note that in the `tsconfig.base.json` we removed the `rootDir` from our TypeScript configuration (compared to the pure TypeScript path aliases setup). The reason is that we no longer treat the entire workspace as a single TypeScript project. Instead, each project's `tsconfig.json` forms its own TypeScript root and will be processed by TypeScript's project references individually.
+
+The `myapp` and `lib-a` configurations look as follows:
+
+{% tabs %}
+
+{% tab label="apps/myapp/tsconfig.json" %}
+
+```json {% fileName="apps/myapp/tsconfig.json" %}
+{
+  "extends": "../../tsconfig.base.json",
+  "compilerOptions": {
+    "outDir": "../../dist/apps/myapp",
+    "rootDir": "src",
+    "tsBuildInfoFile": "../../dist/apps/myapp/tsconfig.tsbuildinfo"
+  },
+  "references": [{ "path": "../../packages/lib-a" }],
+  "include": ["src/**/*"]
+}
+```
+
+{% /tab %}
+
+{% tab label="lib-a TSConfig" %}
+
+```json {% fileName="packages/lib-a/tsconfig.json" %}
+{
+  "extends": "../../tsconfig.base.json",
+  "compilerOptions": {
+    "outDir": "../../dist/packages/lib-a",
+    "rootDir": "src",
+    "tsBuildInfoFile": "../../dist/packages/lib-a/tsconfig.tsbuildinfo"
+  },
+  "include": ["src/**/*"]
+}
+```
+
+{% /tab %}
+{% /tabs %}
+
+Changing the configuration alone isn't enough. If you continue using `tsc` (or `tsc -p`), TypeScript will ignore your project references and treat the workspace as a single large project. To fully leverage project references, you must switch to using the `--build` (`-b`) flag with `tsc`. This mode enables TypeScript to process each project individually, respecting dependencies defined in the references property.
+
+```json {% fileName="package.json" %}
+{
+  "name": "ts-monorepo-linking",
+  ...
+  "scripts": {
+    "dev": "tsx --tsconfig tsconfig.base.json apps/myapp/src/index.ts",
+    "build": "tsc --build",
+    "clean": "tsc --build --clean",
+    "typecheck": "tsc --build --emitDeclarationOnly"
+  }
+}
+```
+
+> Note: `--noEmit` is not compatible with the `--build` flag. Use `--emitDeclarationOnly`.
+
+### Observations
+
+**Dependency resolution:**
```

Comment:
Something weird is happening with this that is not visible in the diff view, there is no newline after dependency resolution in the rendered page

Reply:
oh I see. I'll fix those, thanks